[Framework] iOS style blurring and `ImageFilterConfig`

dkwingsmt opened the pull request:
This PR adds the framework support for a new iOS-style blur. The new style, which I call "bounded blur", works by adding parameters to the blur filter that specify the bounds for the region that the filter sources pixels from.

As discussed in design doc [flutter.dev/go/ios-style-blur-support](http://flutter.dev/go/ios-style-blur-support), it's impossible to pass layout information to filters with the current `ImageFilter` design. Therefore this PR creates a new class `ImageFilterConfig`.

This PR also applies bounded blur to `CupertinoPopupSurface`. The following images show the different looks of a dialog in front of background with abrupt color changes just outside of the border. Notice how the abrupt color changes no longer bleed in.

<img width="639" height="411" alt="image" src="https://github.com/user-attachments/assets/4ceb9620-1056-45c3-b5fa-2ed16d90aace" />

<img width="639" height="411" alt="image" src="https://github.com/user-attachments/assets/abe564f7-ea60-4d07-ad58-063c0e3794a5" />

This feature continues to matter for iOS 26, since the liquid glass design also heavily features blurring.

Fixes https://github.com/flutter/flutter/issues/99691.

### API changes

* `BackdropFilter`: Add `filterConfig`
* `RenderBackdropFilter`: Add `filterConfig`. Deprecate `filter`.
* `ImageFilter`: Add `debugShortDescription` (previously private property `_shortDescription`)

### Demo

The following demo app, which is implemented on the widget layer, compares the effect of a bounded blur and an unbounded blur.

https://github.com/user-attachments/assets/f715db44-c0a0-4ac8-a163-6b859665b032

<details>
<summary>
Demo source
</summary>

```
// Add to pubspec.yaml:
//
//  assets:
//      - assets/kalimba.jpg
//
// and download the image from
// https://github.com/flutter/flutter/blob/ec6f55023760ea4f44d311b9c69c39910f6b8b0c/engine/src/flutter/impeller/fixtures/kalimba.jpg

import 'package:flutter/material.dart';

void main() => runApp(const MaterialApp(home: BlurEditorApp()));

class ControlPoint extends StatefulWidget {
  const ControlPoint({
    super.key,
    required this.position,
    required this.onPanUpdate,
    this.radius = 20.0,
  });

  final Offset position;
  final GestureDragUpdateCallback onPanUpdate;
  final double radius;

  @override
  ControlPointState createState() => ControlPointState();
}

class ControlPointState extends State<ControlPoint> {
  bool isHovering = false;

  @override
  Widget build(BuildContext context) {
    return Positioned(
      left: widget.position.dx - widget.radius,
      top: widget.position.dy - widget.radius,
      child: MouseRegion(
        onEnter: (_) { setState((){ isHovering = true; }); },
        onExit: (_) { setState((){ isHovering = false; }); },
        cursor: SystemMouseCursors.move,
        child: GestureDetector(
          onPanUpdate: widget.onPanUpdate,
          child: AnimatedContainer(
            duration: const Duration(milliseconds: 200),
            width: widget.radius * 2,
            height: widget.radius * 2,
            decoration: BoxDecoration(
              shape: BoxShape.circle,
              color: isHovering
                  ? Colors.white.withValues(alpha: 0.8)
                  : Colors.white.withValues(alpha: 0.4),
              border: Border.all(color: Colors.white, width: 2),
              boxShadow: [
                if (isHovering)
                  BoxShadow(
                    color: Colors.white.withValues(alpha: 0.5),
                    blurRadius: 10,
                    spreadRadius: 2,
                  )
              ],
            ),
            child: const Icon(Icons.drag_indicator, size: 16, color: Colors.black54),
          ),
        ),
      ),
    );
  }
}

class BlurPanel extends StatelessWidget {
  const BlurPanel({super.key, required this.blurRect, required this.bounded});

  final Rect blurRect;
  final bool bounded;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Column(
        children: [
          Expanded(
            child: Stack(
              children: [
                Positioned.fill(
                  child: Image.asset(
                    'assets/kalimba.jpg',
                    fit: BoxFit.cover,
                  ),
                ),
                Positioned.fromRect(
                  rect: blurRect,
                  child: ClipRect(
                      child: BackdropFilter(
                    filterConfig: ImageFilterConfig.blur(
                        sigmaX: 10, sigmaY: 10, bounded: bounded),
                    child: Container(),
                  )),
                ),
              ],
            ),
          ),
          Padding(
            padding: const EdgeInsets.all(8.0),
            child: Text(
              bounded ? 'Bounded Blur' : 'Unbounded Blur',
              style: const TextStyle(fontSize: 20, fontWeight: FontWeight.bold),
            ),
          ),
        ],
      ),
    );
  }
}

class BlurEditorApp extends StatefulWidget {
  const BlurEditorApp({super.key});

  @override
  State<BlurEditorApp> createState() => _BlurEditorAppState();
}

class _BlurEditorAppState extends State<BlurEditorApp> {
  Offset p1 = const Offset(100, 100);
  Offset p2 = const Offset(300, 300);

  @override
  Widget build(BuildContext context) {
    final blurRect = Rect.fromPoints(p1, p2);

    return Scaffold(
      body: Stack(
        children: [
          Positioned.fill(
            child: Row(
              children: [
                Expanded(
                  child: BlurPanel(blurRect: blurRect, bounded: true),
                ),
                Expanded(
                  child: BlurPanel(blurRect: blurRect, bounded: false),
                ),
              ],
            ),
          ),

          ControlPoint(position: p1, onPanUpdate: (details) { setState(() => p1 = details.globalPosition); }),
          ControlPoint(position: p2, onPanUpdate: (details) { setState(() => p2 = details.globalPosition); }),
        ],
      ),
    );
  }
}

```

</details>


## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

dkwingsmt committed:
Impl

dkwingsmt committed:
config doc

dkwingsmt committed:
dialog bounded true

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ios-blur-framework

chunhtai commented on packages/flutter/lib/src/widgets/basic.dart:
```diff
@@ -653,18 +656,35 @@ class BackdropFilter extends SingleChildRenderObjectWidget {
   const BackdropFilter.grouped({
     super.key,
-    required this.filter,
+    this.filter,
```
same, consider deprecation and reimplementing with ImageFilterConfig.filter

dkwingsmt replied:
(Discussing this issue in https://github.com/flutter/flutter/pull/175473#discussion_r2377012519)

chunhtai commented on packages/flutter/lib/src/rendering/image_filter_config.dart:
```diff
@@ -0,0 +1,182 @@
+
+/// A configuration for a [ui.ImageFilter].
+///
+/// This class provides a framework-level abstraction for image filters that can
```
Just curious have you thought about renaming existing `ImageFilter` to something else and reuse the name for this class?

dkwingsmt replied:
That's quite a clever duck typing. Well I can think of one way it can break: `ImageFilter` actually comes from `dart:ui` and people might write import `dart:ui` as `ui` and write `ui.ImageFilter` instead.

chunhtai replied:
That won't be a problem if this is definded in dart:ui right? though that means we need to define both web and not web version of this

dkwingsmt commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1201,11 +1204,15 @@ class RenderBackdropFilter extends RenderProxyBox {
   RenderBackdropFilter({
     RenderBox? child,
-    required ui.ImageFilter filter,
+    ui.ImageFilter? filter,
```
For now I don't have any plan to deprecate this. `ImageFilterConfig` is a bit more tedious to use than `ImageFilter` and its only benefit is the iOS-style bounded blur. Forcing everyone to migrate to that isn't worth it.

dkwingsmt committed:
Comments

dkwingsmt commented:
Just to be clear, the obstacle of the getter has nothing to do with resolving.

If we only store `_filterConfig` in the class, then the `get filter` will need to somehow extract the `ImageFilter` object that was assigned via either the constructor or the setter. But the `ImageFilterConfig` class does not have a way to query its direct filter (when applicable). Therefore we can't write a `get filter`, which is further required for the setter. In this way, the `filter` property can not be queried or set, breaking all code that uses it. I doubt anyone uses the render object though, so maybe we can just delete it as well.

dkwingsmt commented on packages/flutter/lib/src/rendering/image_filter_config.dart:
```diff
@@ -0,0 +1,182 @@
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType) {
```
As far as I see `if (other.runtimeType != runtimeType)` is the convention we used for many, if not all, `bool operator ==` implementations. 

(During this check I found that our convention also includes an `identical` check, which I've added to the PR.)

dkwingsmt commented on packages/flutter/lib/src/rendering/image_filter_config.dart:
```diff
@@ -0,0 +1,182 @@
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    return other is _DirectImageFilterConfig &&
```
As far as I see `if (other.runtimeType != runtimeType)` is the convention we used for many, if not all, `bool operator ==` implementations. 

(During this check I found that our convention also includes an `identical` check, which I've added to the PR.)

dkwingsmt committed:
Remove render object filter

dkwingsmt commented:
I've removed the `filter` property from the render object while keeping it in the widget. The reasons are:
* The render object should not be used much and hence safe to break.
* The widget is commonly used and should not be broken.
* The widget is immutable and having two properties there is much less of an issue.

Let me know what you think!

chunhtai commented on packages/flutter/lib/src/rendering/image_filter_config.dart:
```diff
@@ -0,0 +1,174 @@
+
+import 'dart:ui' as ui;
+
+import 'package:flutter/foundation.dart';
```
what is this used for?

chunhtai commented on packages/flutter/lib/src/rendering/image_filter_config.dart:
```diff
@@ -0,0 +1,174 @@
+  /// Resolves this configuration into a [ui.ImageFilter], given the
+  /// `bounds` of the widget applying the filter.
+  ///
+  /// The `bounds` can be used to create layout-dependent filters, such as
```
somewhere in the doc should mention what coordinate this bound is in. based on the code, This is in global coordinate

dkwingsmt replied:
Fixed.

dkwingsmt replied:
Update: Since the parameter here has been changed to a `ImageFilterContext`, it no longer need to define the bounds. Instead, `ImageFilterContext.bounds` contains a complete definition of the bounds, including the coordinate it uses.

chunhtai commented on packages/flutter/lib/src/rendering/proxy_box.dart:
```diff
@@ -1271,17 +1273,37 @@ class RenderBackdropFilter extends RenderProxyBox {
+  // transform is applied to the rectangle that represents this object's size.
+  Rect _sizeForFilter(Offset offset) {
+    final Matrix4 transform = Matrix4.translationValues(-offset.dx, -offset.dy, 0);
+    for (RenderObject current = this; current.parent != null; current = current.parent!) {
```
you can use getTransformTo to get the global transform

dkwingsmt replied:
If I remember correctly I've tried it, but it doesn't include the root transform of `RenderView`. Rather than combining `getTransformTo` with the root transform, I think it's easier to compute everything this way.

chunhtai reviewed (commented):
Approach overall looks good, just wonder if we can do anything about https://github.com/flutter/flutter/pull/175473/files#r2376931055

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ios-blur-framework

dkwingsmt committed:
Add config class

dkwingsmt commented:
I added a `ImageFilterContext` class that groups the data to resolve filters, so that we can expand the parameter set without breakages. I also added documents. Let me know what you think. @chunhtai

Piinks commented:
Greetings from stale PR triage! 👋 
Is this change still on your radar?

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ios-blur-framework

dkwingsmt committed:
toString; doc; fix implementation

dkwingsmt committed:
Doc

dkwingsmt committed:
Doc

dkwingsmt committed:
unit tests and shortDescription

dkwingsmt committed:
Add .filter getter and unit test

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ios-blur-framework

dkwingsmt committed:
Rename to staticFilter; analyze

dkwingsmt committed:
Rename back to filter

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ios-blur-framework

dkwingsmt committed:
Change deprecation version

dkwingsmt committed:
apply shortDescription to more

dkwingsmt committed:
Fix _boundsString

dkwingsmt committed:
Fix build

dkwingsmt committed:
remove shortdesc in some class

dkwingsmt committed:
.1, and fix description

dkwingsmt committed:
Skip on web

dkwingsmt committed:
Fix compile

dkwingsmt committed:
Fix tests

dkwingsmt committed:
Merge branch 'master' into ios-blur-framework

chunhtai commented on packages/flutter/lib/src/rendering/image_filter_config.dart:
```diff
@@ -0,0 +1,258 @@
+///  * [BackdropFilter.filterConfig], which uses this class to configure its effect.
+@immutable
+abstract class ImageFilterConfig {
+  /// Creates a configuration that directly uses the given filter.
```
Maybe mention this won't resolve the given image filter with the rect if they want to use ImageFilterConfig(ImageFilter.blur), they should use ImageFilterConfig.blur

chunhtai reviewed (commented):
LGTM for the most part, just some minor comments

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ios-blur-framework

dkwingsmt committed:
Rename to debugShortDescription

dkwingsmt committed:
Doc

dkwingsmt committed:
Fix overriden function

dkwingsmt committed:
Merge branch 'ios-blur-framework' of github.com:dkwingsmt/flutter into ios-blur-framework

dkwingsmt commented:
@chunhtai I've addressed all your comments. PTAL. Thank you!

chunhtai reviewed (approved):
LGTM, except for the a doc nit

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into ios-blur-framework

dkwingsmt committed:
Fix doc
